crates/sui-network/src/state_sync/mod.rs

```diff
@@ -0,0 +1,854 @@
+                self.store.clone(),
+                self.peer_heights.clone(),
+                self.config.checkpoint_header_download_concurrency(),
+                // The if condition should ensure that this is Some
+                highest_known_checkpoint.unwrap(),
+            )
+            .map(|result| match result {
+                Ok(()) => {}
+                Err(e) => {
+                    debug!("error syncing checkpoint {e}");
+                }
+            });
+            let task_handle = self.tasks.spawn(task);
+            self.sync_checkpoint_summaries_task = Some(task_handle);
+        }
+    }
+
+    fn maybe_start_checkpoint_contents_sync_task(&mut self) {
+        // Only run one sync task at a time
+        if self.sync_checkpoint_contents_task.is_some() {
+            return;
+        }
+
+        let highest_trusted_checkpoint = self.store.get_highest_trusted_checkpoint();
+        let highest_synced_checkpoint = self.store.get_highest_synced_checkpoint();
+
+        if highest_trusted_checkpoint
+            .as_ref()
+            .map(|x| x.sequence_number())
+            > highest_synced_checkpoint
+                .as_ref()
+                .map(|x| x.sequence_number())
+        {
+            let task = sync_checkpoint_contents(
+                self.network.clone(),
+                self.store.clone(),
+                self.peer_heights.clone(),
+                self.weak_sender.clone(),
+                self.checkpoint_event_sender.clone(),
+                self.config.transaction_download_concurrency(),
+                // The if condition should ensure that this is Some
+                highest_trusted_checkpoint.unwrap(),
+            );
+
+            let task_handle = self.tasks.spawn(task);
+            self.sync_checkpoint_contents_task = Some(task_handle);
+        }
+    }
+
+    fn spawn_notify_peers_of_checkpoint(&mut self, checkpoint: VerifiedCheckpoint) {
+        let task =
+            notify_peers_of_checkpoint(self.network.clone(), self.peer_heights.clone(), checkpoint);
+        self.tasks.spawn(task);
+    }
+}
+
+async fn notify_peers_of_checkpoint(
+    network: anemo::Network,
+    peer_heights: Arc<RwLock<PeerHeights>>,
+    checkpoint: VerifiedCheckpoint,
+) {
+    let futs = peer_heights
+        .read()
+        .unwrap()
+        .heights
+        .iter()
+        // Filter out any peers who we know already have a checkpoint higher than this one
+        .filter(|(_peer_id, &height)| Some(checkpoint.sequence_number()) > height)
+        .map(|(peer_id, _height)| peer_id)
+        // Filter out any peers who we aren't connected with
+        .flat_map(|peer_id| network.peer(*peer_id))
+        .map(StateSyncClient::new)
+        .map(|mut client| {
+            let request = Request::new(checkpoint.inner().clone()).with_timeout(DEFAULT_TIMEOUT);
+            async move { client.push_checkpoint_summary(request).await }
+        })
+        .collect::<Vec<_>>();
```

Comment:
the collect isn't necessary right? I think you could pass an uncollected iterator to join_all?

Reply:
Yes you can pass an iterator to join_all but in this case we probably don't want to do that. The iterator is created by holding a read-lock on the peer_heights value and that can't be held across an await point which may happen depending on the internals of join_all. So instead we explicitly create and collect all the futures we want, release the read-lock and then begin awaiting on the futures.

Reply:
Thanks, I totally missed the read lock! One minor bikeshedding thing, feel free to ignore if you hate it, but

everywhere you have
 let foo = bar
        ....
        .collect::<SomeType>();

I'd personally find it easier to read as

let foo: SomeType = bar
        ....
        .collect();